### ISSUE TITLE
net.Read/WriteAngle/Vector not precise compared to lua

### ISSUE THREAD

**Issue by meepdarknessmeep** [Facepunch/garrysmod-issues#1947]
compared to

**Comment by thegrb93** [comment]
Lua uses double precision and floats are single precision.

**Comment by meepdarknessmeep** [comment]
No. I use net.WriteFloat and it's more precise than the c version.

**Comment by thegrb93** [comment]
Oh so those functions aren't as precise as WriteFloat. Maybe source engine floats are 16 bit?

**Comment by meepdarknessmeep** [comment]
I highly doubt that.

**Comment by robotboy655** [comment]
That's how source engine networks things, I am fairly certain this is the game will ALL networked vectors.

**Comment by meepdarknessmeep** [comment]
The problem is that it doesn't lose precision in lua for some reason. Is there any way it could be fixed?

**Comment by robotboy655** [comment]
I am not sure what you mean

**Comment by thegrb93** [comment]
He means WriteVector/WriteAngle should be made single precision. Idk though, the pros is better precision, but idk if the performance hit is worth it.

**Comment by robotboy655** [comment]
They use engine functions to network stuff, there are no precision changes or w/e.

**Comment by thegrb93** [comment]
He says the floats on their own are more precise than angles or vectors though.

**Comment by robotboy655** [comment]
Yes, floats are networked as floats, vectors are networked as vectors. With engine functions.

**Comment by meepdarknessmeep** [comment]
Could you at least change the precision of vectors in the lua api for garrys mod?
It kind of bothers me that it does this.

**Comment by robotboy655** [comment]
Lua vectors use Valves Vector class, which uses floats as members.

**Comment by meepdarknessmeep** [comment]
Oops sorry, long day, 
I meant change the net.Read/WriteVector functions to write three floats instead of whatever valve does in their c function to send them.

**Comment by Doliman100** [comment]
If anyone is interested in how this works, `net.WriteVector` just calls `bf_write::WriteBitVec3Coord` from the Source SDK which sends data as fixed-point rather than floating-point. It uses something similar to the Q14.5 format. Only 5 fractional bits are sent. In decimal it's only 1 digit of precision log<sub>10</sub>(2<sup>5</sup>) ≈ 1.505.
https://github.com/ValveSoftware/source-sdk-2013/blob/c623a7c30d5cb7275cc64ed0b866f61f4a64c6eb/src/tier1/bitbuf.cpp#L654
https://en.wikipedia.org/wiki/Q_(number_format)
Some other issues related to this function: https://github.com/wiremod/wire/issues/240 https://github.com/wiremod/wire/issues/93
